### DISCUSSION TITLE
Is it possible to add usage relations to the collaboration diagrams for properties?

### DISCUSSION THREAD

**Discussion by ahayzen-kdab** [doxygen/doxygen#11007] [Q&A]
With the following code, I would like to show a relation from `Main` to `Other` as it is used in a property.

```cpp
class Other : public QObject
{
  Q_OBJECT

public:
  void publicMethod();

private:
  void privateMethod();

  int mPrivateMember;
};

class Main : public QObject
{
  Q_OBJECT

  Q_PROPERTY(Other* other READ other)

public:
  Other* other() const;

private:
  void privateMethod();

  Other* mOther;
};
```

If I run doxygen with the following config with `EXTRACT_PRIVATE` enabled then I get a nice UML diagram as shown below

```
EXTRACT_ALL            = YES
HAVE_DOT               = YES
UML_LOOK               = YES
EXTRACT_PRIVATE = YES
```

![image](https://github.com/user-attachments/assets/c3aa0032-8948-472c-88e6-2e172022091f)

If I now turn off `EXTRACT_PRIVATE` the relation is lost.

![image](https://github.com/user-attachments/assets/b0cac438-94f7-4265-bfe0-4ba41177b424)

I would like to be able to generate a collaboration diagram that includes usage relations but does not show the private members and methods. It appears that members of a class are being treated in a special way to create these relations but properties are not ? (even though here the property and member are basically the same thing)

Is there a way to either
- Allow for properties to be considered in a similar way to members so that they cause a usage relation? (even if it's behind a config option)
- A way to manually specify a relation in the documentation ? eg so i could put `\links Other` or `\usagerelation Other` etc on the `Main` class or property somewhere to cause a relation

**Comment by albert-github** [comment]
- Which version of doxygen are you using?
- So you would like to see something like:
  ![image](https://github.com/user-attachments/assets/d96629e9-e855-47d0-b571-5cea50f80abc)
  but without the `mOther`.
  (diagram created by making `mOther` protected)

**Reply by ahayzen-kdab** [reply]
I'm currently using version `1.9.1` from Ubuntu 22.04, but can use a newer version from source etc if required.

And yes, ideally the diagram would look like that !

My initial hope had been that doxygen would treat properties in a similar way to members and consider those as a usage relation, but i have not found a config option to do this.

So then I thought maybe an explicit relation could be made in the documentation somewhere, but i cannot see a custom command that would allow me to create a relation.

**Reply by ahayzen-kdab** [reply]
The `protected` workaround is curious and might be good enough, so that we can see public methods, can see relations, but don't see private members/methods. Maybe I can make a macro/define so that these members are only protected when building docs :thinking:

**Reply by albert-github** [reply]
> Maybe I can make a macro/define so that these members are only protected when building docs

Is dangerous as it implies something that is not true.

**Comment by albert-github** [comment]
The current doxygen version is 1.11.0, the version 1.9.1 is from January 8, 2021 so an update doesn't hurt.

Using `DOT_UML_DETAILS = NONE` I get
![image](https://github.com/user-attachments/assets/fa4ade93-2f76-4351-8d00-b5523ac89d48)
but we still have here `-mOther` (@doxygen is this logical or should this be removed as well, maybe by another possibility of `DOT_UML_DETAILS`)

**Reply by albert-github** [reply]
Regarding the idea "maybe by another possibility of `DOT_UML_DETAILS`" I've just pushed a proposed patch, pull request #11008
